src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,276 @@
+
+    private List<String> readBucket(long bucket) {
+      Timer.Context timer = readBucketTimer.time();
+      try {
+        return accountsManager.getNumbersInBucket((1L << BUCKET_ORDER) - 1L, bucket);
+      } finally {
+        timer.stop();
+      }
+    }
+
+    private void sendBucket(long bucket, List<String> numbers) {
+      Timer.Context timer = sendBucketTimer.time();
+      try {
+        Response response = client.target(serverApiUrl)
+                                  .path(String.format("/v1/directory/reconcile/%d/%d", BUCKET_COUNT, bucket))
+                                  .request(MediaType.APPLICATION_JSON)
+                                  .put(Entity.json(new DirectoryReconciliationRequest(numbers)));
+        if (response.getStatus() != 200) {
+          sendBucketErrorMeter.mark();
+          logger.warn("http error " + response.getStatus());
+        }
+      } catch (ProcessingException ex) {
+        sendBucketErrorMeter.mark();
+        logger.warn("request error: ", ex);
+      } finally {
+        timer.stop();
+      }
+    }
+
+    @Override
+    public void run() {
+      logger.info("Directory reconciliation worker " + workerId + " started");
+
+      try (Jedis jedis = jedisPool.getWriteResource()) {
+        long nowMs = System.currentTimeMillis();
+        jedis.zremrangeByScore(WORKERS_KEY, Double.NEGATIVE_INFINITY, (double) (nowMs - WORKER_TTL_MS));
+      }
+
+      long workIntervalMs = 0;
+      long lastWorkTimeMs = 0;
+      long lastPingTimeMs = 0;
+      long randomJitterMs = 0;
+      while (sleepWhileRunning(Math.min(lastWorkTimeMs + workIntervalMs + randomJitterMs,
+                                        lastPingTimeMs + WORKER_TTL_MS / 2) - System.currentTimeMillis())) {
+        long nowMs = System.currentTimeMillis();
+
+        try (Jedis jedis = jedisPool.getWriteResource()) {
+          lastPingTimeMs = nowMs;
+
+          jedis.zadd(WORKERS_KEY, (double) nowMs, workerId);
+
+          if (nowMs - lastWorkTimeMs > workIntervalMs + randomJitterMs) {
+            lastWorkTimeMs = nowMs;
+
+            long workerCount = jedis.zcount(WORKERS_KEY, (double) (nowMs - WORKER_TTL_MS), Double.POSITIVE_INFINITY);
+
+            workIntervalMs = 86400_000L * workerCount / BUCKET_COUNT;
+            randomJitterMs = (long) (random.nextDouble() * JITTER_MAX * workIntervalMs);
+
+            long counter = jedis.bitfield(COUNTER_KEY, COUNTER_ARGS).get(0);
+            long bucket  = counter % BUCKET_COUNT;
+
+            List<String> numbers = readBucket(bucket);
+            sendBucket(bucket, numbers);
+          } else if (lastWorkTimeMs > nowMs) {
+            lastWorkTimeMs = nowMs;
+          }
+        }
+      }
+
+      try (Jedis jedis = jedisPool.getWriteResource()) {
+        jedis.zrem(WORKERS_KEY, workerId);
+      }
+
+      logger.info("Directory reconciliation worker " + workerId + " shut down");
+
+      synchronized (this) {
+        finished = true;
+        notifyAll();
+      }
```

Comment:
It seems like the above is going to have to change with the change in "bucket" strategy, so I'll wait to review

Reply:
revised in 2d295f73c5cd75724fb709f2cf5dfdaca4f6573e